controllers/main.py

```diff
@@ -0,0 +1,183 @@
+        self.y = y
+        self.width = 10  # temporary when we don't have real images
+        self.height = 10  # must be height of real image
+        self.bonus = bonus
+        self.speed = speed
+        self.is_dead = False
+        self.bullet = Bullet(self, bullet_filename)
+        self.shift = 5
+
+    def move_to(self, x, y):
+        self.x = x
+        self.y = y
+
+    def check_fire(self, other_unit):
+        raise NotImplementedError
+
+    def fire(self, other_unit):
+        raise NotImplementedError
+
+    def make_bullet(self):
+        raise NotImplementedError
+
+
+class Invader(Unit):
+    moving_speed = 0
+
+    def __init__(self, hero_filename, bullet_filename, x, y, bonus=0, speed=5):
+        super(Invader, self).__init__(hero_filename, bullet_filename, x, y, bonus, speed)
+        self.moving_speed = speed
+
+    def check_if_move_y(self):
+        return self.x <= self.shift
+
+    def set_speed(self, game_field_width):
+        # this check for first Invader
+        if self.x <= self.shift:
+            self.moving_speed = self.speed or -self.speed  # positive value
+        # this check for last Invader
+        if self.x >= game_field_width:
+            self.moving_speed = self.speed < 0 and self.speed or -self.speed  # negative value
+
+    def check_fire(self, other_unit):
+        raise NotImplementedError
+
+    def fire(self, hero):
+        hero.is_dead = True
+
+    def make_bullet(self):
+        # Invaders will be at the top and must bullet at the bottom
+        # so speed must be positive value
+        self.bullet.move(self.speed or -self.speed)
+
+
+class Hero(Unit):
+
+    def __init__(self, unit_filename, bullet_filename, x, y, bonus=0, speed=5, life_count=3):
+        super(Hero, self).__init__(unit_filename, bullet_filename, x, y, bonus, speed)
+        self.life_count = life_count
+
+    def fire(self, invader):
+        if self.life_count > 0:
+            self.life_count -= 1
+            invader.is_dead = True
+            self.bonus += invader.bonus
+        else:
+            self.is_dead = True
+
+    def make_bullet(self):
+        # hero will be at the bottom and must bullet at the top
+        # so speed must be negative value
+        self.bullet.move(-self.speed)
+
+    def check_fire(self, invader):
+        # check if coordinate the bullet and the Invader's is the same
+        if (self.bullet.x > invader.x - invader.width) and (self.bullet.x < invader.x + invader.width) and \
+                (self.bullet.y > invader.y - invader.height) and (self.bullet.y < invader.y + invader.height):
+            self.fire(invader)
+            return True
+        elif self.bullet.y < 0:
+            return True
+        else:
+            return False
+
+
+class Bullet():
+    def __init__(self, unit, image):
+        self.x = unit.x
+        self.y = unit.y
+        self.image = image
+
+    def move(self, moving_speed):
+        self.y += moving_speed
+
+
+class GameController(object):
```

Comment:
In future of Game Controller I'd like to see (possibly inherited class of it) a method run which exactly starts the game. On this run we will be reacting on events. The controller must control coordinates of each Unit and collisions with other objects. For example Invider is killed once reached a bullet to its coordinates, or decreased live for a Hero when also colides to Bullet. Hero can be also injured by direct collision with an Invider or any other object.

Game Controller must store map of coordinates for each object just to send it to clients. Also it will run AsyncLoop for processing all events. For example when we do Fire, a bullet will initiate its run and will be running in a loop until reaches final destination. On each loop of the bullet (the same rule actually works for other object) we have to be able triggering event to GameController and let it check for collisions with other objects (for such checks we can also implement some kind of Game Rules object which can be used here by Game Controller)